ext/oj/dump_compat.c

```diff
@@ -766,6 +799,15 @@ dump_obj(VALUE obj, int depth, Out out, bool as_ok) {
 	oj_dump_raw_json(obj, depth, out);
 	return;
     }
+
+    if (rb_cTime == rb_obj_class(obj)) {
+        if (as_ok && rb_respond_to(obj, oj_as_json_id)) {
+            dump_as_json(obj, depth, out, true);
+        } else {
+            oj_dump_ruby_time(obj, out);
+        }
+        return;
+    }
```

Comment:
If my reading of this is correct, `Time.to_json` will never be called. Is that the intent and if so I would expect there will be issues raised it these changes are released,

Reply:
Exactly, you're right. To improve performance, like a `rails` mode, I want to avoid Time#to_json calling with this PR.

Reply:
The problem with that approach is that if someone defines a `#to_json` they will expect it to be called. With this change it will not be called yet with the JSON it would so Oj in compat mode would not longer be compatible with the JSON gem.

Reply:
Yeah, I have same concerns from the beginning, so this patch e focus on Time class only.

Reply:
I think anyone does not override Time#to_json method.

Reply:
So maybe a check that the dump is in rails mode is needed or possibly just make sure the rails dump code is patched by not the compat code.

Reply:
If we use `method` and `super_method`, we may be able to know if `to_json` was overridden.

Here is a sample code.

```ruby
class Foo
  def say(arg)
    puts arg
  end
end

class Bar < Foo
  def say(arg)
    super
  end
end

p Bar.new.method(:say).super_method # => #<Method: Foo#say>
p Foo.new.method(:say).super_method # => nil
```

Reply:
If the goal is to just improve the Rails handling of time then I think it would be better to look at the rails.c file. In particular, the `dump_obj()` function near the bottom of the file. If in rails mode all the type including time should have an `#as_json` method which should be called in preference to the `#to_json`. 

As for not overriding the `#to_json` method on time it is hard to know what people will do. It has happened more than once that I changed or removed something I though was never used only to have an issue created by some use that was using what I expected could be changed. We don't have a very good feedback cycle with gem feature use I'm afraid.

Reply:
This patch already calls #as_json in preference to #to_json...

hmm, am I mistaken about something?

Reply:
The patch checks for `#as_json` but if that is not present it drops back to the calling `oj_dump_ruby_time()` which just calls `#to_s` on Time. No chance to run the `#to_json` so it is not that there is a preference for `#as_json` over `#to_json` but rather it is either `#as_json` or nothing (well `#to_s`).

Reply:
Maybe it would help if the use case were more clear. I think you are trying to solve the issue for Rails but not for direct JSON gem use, right?

Reply:
If I am missing something I would be glad to get on a call with you. Zoom of something.

Reply:
The main motivation:

I hope to use only Oj gem in my rails app. However, the gems which exists somewhere in dependency requires JSON gem. And it decreases a Time#to_json performance. I want to solve my problem

Reply:
I see. It sounds like the best option would be to see if the `:rails` mode is in use. If it is in your tests then it would be okay to not check for `#to_json` I think. Can you check? That might be the way to get around the slow `#to_json`.

Reply:
The JSON API in my rails app uses oj with :compat mode ... hmm.

Reply:
That's a tough one. That implies the author wanted JSON gem behaviour of calling `#to_json`. That or didn't know any better. Is it your code or can it be monkey patched?

Reply:
It seems difficult to merge with my little spare time, so close it.

Reply:
Sorry to hear that.

Reply:
Sorry to spend your time and I hope that you have an overview of this problem.

I'll try again when I have more time :)